rust-runtime/aws-smithy-checksums/src/body.rs

```diff
@@ -0,0 +1,342 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use crate::http::{new_from_algorithm, HttpChecksum};
+
+use aws_smithy_http::body::SdkBody;
+use aws_smithy_http::header::append_merge_header_maps;
+
+use bytes::Bytes;
+use http::{HeaderMap, HeaderValue};
+use http_body::SizeHint;
+use pin_project_lite::pin_project;
+
+use std::fmt::Display;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+pin_project! {
+    /// A `ChecksumBody` will read and calculate a request body as it's being sent. Once the body has
+    /// been completely read, it'll append a trailer with the calculated checksum.
+    pub struct ChecksumBody<InnerBody> {
+            #[pin]
+            body: InnerBody,
+            checksum: Option<Box<dyn HttpChecksum>>,
+    }
+}
+
+impl ChecksumBody<SdkBody> {
+    /// Given an `SdkBody` and a `Box<dyn HttpChecksum>`, create a new `ChecksumBody<SdkBody>`.
+    pub fn new(body: SdkBody, checksum: Box<dyn HttpChecksum>) -> Self {
+        Self {
+            body,
+            checksum: Some(checksum),
+        }
+    }
+}
+
+impl http_body::Body for ChecksumBody<SdkBody> {
+    type Data = Bytes;
+    type Error = aws_smithy_http::body::Error;
+
+    fn poll_data(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Option<Result<Self::Data, Self::Error>>> {
+        let this = self.project();
+        match this.checksum {
+            Some(checksum) => {
+                let poll_res = this.body.poll_data(cx);
+                match &poll_res {
+                    Poll::Ready(Some(Ok(data))) => checksum.update(&data),
+                    _ => (),
+                }
+
+                poll_res
+            }
+            _ => Poll::Ready(Some(Err(Box::new(Error::Taken)))),
+        }
+    }
+
+    fn poll_trailers(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<Result<Option<HeaderMap>, Self::Error>> {
+        let this = self.project();
+        let poll_res = this.body.poll_trailers(cx);
+
+        if let Poll::Ready(Ok(maybe_inner_trailers)) = poll_res {
+            let checksum_headers = if let Some(checksum) = this.checksum.take() {
+                checksum.headers()
+            } else {
+                return Poll::Ready(Err(Box::new(Error::Taken)));
+            };
+
+            return match maybe_inner_trailers {
+                Some(inner_trailers) => Poll::Ready(Ok(Some(append_merge_header_maps(
+                    inner_trailers,
+                    checksum_headers,
+                )))),
+                None => Poll::Ready(Ok(Some(checksum_headers))),
+            };
+        }
+
+        poll_res
+    }
+
+    fn is_end_stream(&self) -> bool {
+        self.body.is_end_stream()
```

Comment:
this is actually wrong and **may** cause us to not send trailers:
From the docs:
```
An end of stream means that both poll_data and poll_trailers will return None.
```

Reply:
I fixed this by checking if the inner `HttpChecksum` has been taken. That only happens once we're reading trailers and the inner body has emitted any trailers it contains.